--- a/drivers/net/phy/realtek.c	2025-01-13 09:09:27.299820256 -0600
+++ b/drivers/net/phy/realtek.c	2025-01-19 11:39:44.650454453 -0600
@@ -925,2 +925,4 @@ static int rtl822x_get_features(struct p
 
+	linkmode_set_bit(ETHTOOL_LINK_MODE_Autoneg_BIT, phydev->supported);
+
 	linkmode_mod_bit(ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
@@ -938,2 +940,4 @@ static int rtl822x_config_aneg(struct ph
 	int ret = 0;
+	u32 v;
+		phydev_warn(phydev, "%s: Sri autoneg:%d\n", __func__, phydev->autoneg);
 
@@ -941,2 +945,25 @@ static int rtl822x_config_aneg(struct ph
 		u16 adv = linkmode_adv_to_mii_10gbt_adv_t(phydev->advertising);
+		phydev_warn(phydev, "%s: adv:%d\n", __func__, adv);
+
+//		u32 v = phy_read_mmd(phydev, MDIO_MMD_VEND1, 0x7582);
+//		phydev_warn(phydev, "%s: SRI adding 0x7582 before %x\n", __func__, v);
+//		v = v|0x11E1;
+//		phydev_warn(phydev, "%s: SRI adding 0x7582 after %x\n", __func__, v);
+//		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1, 0x7582, v);
+//		phydev_warn(phydev, "%s: SRI ret value %d\n", __func__, ret);
+
+		v = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_ADVERTISE);
+		v |= ADVERTISE_10HALF;
+		v |= ADVERTISE_10FULL;
+		v |= ADVERTISE_100HALF;
+		v |= ADVERTISE_100FULL;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_ADVERTISE, v);
+		/* Allow 1GBit */
+		v = phy_read_mmd(phydev, MDIO_MMD_VEND2, 0xA412);
+		v |= ADVERTISE_1000FULL;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND2, 0xA412, v);
+		/* Allow 2.5G */
+		v = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_10GBT_CTRL);
+		v |= MDIO_AN_10GBT_CTRL_ADV2_5G;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_10GBT_CTRL, v);		
 
@@ -946,6 +973,24 @@ static int rtl822x_config_aneg(struct ph
 					       adv);
-		if (ret < 0)
-			return ret;
-	}
+		// Perform software reset and turn on auto-neg
+		//phy_write_mmd(phydev, MDIO_MMD_AN, 0x0, 0x1200); //bit9-restart an, bit12-enable an, bit15-soft reset
 
+				/* AutoNegotiationEnable */
+		v = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_CTRL1);
+		v |= MDIO_AN_CTRL1_ENABLE; /* Enable AN */
+		//v &= ~MDIO_AN_CTRL1_ENABLE; //Disable AN
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_CTRL1, v);
+	
+		//v = phy_read_mmd(phydev, 0x1, 0x0); //PMA PMD SPEED CONTROL
+		//v =0x2058; // Force 2.5g speed
+		//ret = phy_write_mmd(phydev, 0x1, 0x0, v);	
+		/* RestartAutoNegotiation */
+		v = phy_read_mmd(phydev, MDIO_MMD_VEND2, 0xA400);
+		v |= MDIO_AN_CTRL1_RESTART;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND2, 0xA400, v);
+		if (ret < 0){
+			phydev_warn(phydev, "%s: SRI ret value %d\n", __func__, ret);
+		}
+		return 0;
+	}
+	phydev_warn(phydev, "%s: Sri auto neg not enabled :%d\n", __func__, phydev->autoneg);
 	return __genphy_config_aneg(phydev, ret);
@@ -978,2 +1023,3 @@ static int rtl822x_read_status(struct ph
 	int lpadv, ret;
+		phydev_warn(phydev, "%s: Sri autoneg:%d\n", __func__, phydev->autoneg);
 
@@ -985,9 +1031,15 @@ static int rtl822x_read_status(struct ph
 	    !phydev->autoneg_complete) {
-		mii_10gbt_stat_mod_linkmode_lpa_t(phydev->lp_advertising, 0);
-		return 0;
+					phydev_warn(phydev, "%s: Sri autoneg complete:%d\n", __func__, phydev->autoneg_complete);
+		//mii_10gbt_stat_mod_linkmode_lpa_t(phydev->lp_advertising, 0);
+		//return 0;
 	}
 
-	lpadv = phy_read_paged(phydev, 0xa5d, 0x13);
+	lpadv = phy_read_mmd(phydev, MDIO_MMD_VEND1, 0xa5b4); // force speed status register
+	phydev_warn(phydev, "%s: Sri force speed status:%x\n", __func__, lpadv);
+
+	//lpadv = phy_read_paged(phydev, 0xa5d, 0x13);
+	lpadv = phy_read_mmd(phydev, MDIO_MMD_VEND2, 0xa434);
 	if (lpadv < 0)
 		return lpadv;
+	phydev_warn(phydev, "%s: Sri lpadv:%x\n", __func__, lpadv);
 
@@ -1023,2 +1075,3 @@ static int rtl822x_c45_config_aneg(struc
 	int ret, val;
+	phydev_warn(phydev, "%s: Sri autoneg:%d\n", __func__, phydev->autoneg);
 
@@ -1049,2 +1102,3 @@ static int rtl822x_c45_read_status(struc
 	int ret, val;
+		phydev_warn(phydev, "%s: Sri autoneg:%d\n", __func__, phydev->autoneg);
 
@@ -1075,3 +1129,3 @@ static int rtl822x_c45_read_status(struc
 		return val;
-
+	phydev_warn(phydev, "%s: Sri rtl_vndr2_physr :%d\n", __func__, val);
 	rtlgen_decode_physr(phydev, val);
@@ -1103,2 +1157,3 @@ static bool rtlgen_supports_2_5gbps(stru
 	mutex_unlock(&phydev->mdio.bus->mdio_lock);
+		phydev_warn(phydev, "%s: Sri supports 2.5g:%x\n", __func__, (val > 0 && val & MDIO_PMA_SPEED_2_5G));
 
@@ -1185,3 +1240,3 @@ static int rtl822x_probe(struct phy_devi
 
-	val = phy_read_mmd(phydev, MDIO_MMD_VEND1, RTL8221B_PHYCR1);
+	val = phy_read_mmd(phydev, MDIO_MMD_VEND2, RTL8221B_PHYCR1);
 	if (val < 0)
@@ -1194,3 +1249,3 @@ static int rtl822x_probe(struct phy_devi
 
-	phy_write_mmd(phydev, MDIO_MMD_VEND1, RTL8221B_PHYCR1, val);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, RTL8221B_PHYCR1, val);
 
@@ -1523,3 +1578,5 @@ static struct phy_driver realtek_drvs[]
 		.config_aneg    = rtl822x_config_aneg,
-		.read_status    = rtl822x_read_status,
+		.config_init    = rtl822xb_config_init,
+		.get_rate_matching = rtl822xb_get_rate_matching,
+		.read_status	= rtl822xb_read_status,
 		.suspend        = genphy_suspend,
